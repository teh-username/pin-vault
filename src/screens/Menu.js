import React from 'react';
import { connect } from 'react-redux';
import { Alert, ScrollView } from 'react-native';

import {
  getPasscodeRequirement,
  getCurrentPasscode,
  togglePasscodeRequirement,
  resetPasscodeSettings,
} from '../redux/modules/settings';

import Settings from '../components/Settings';
import Debuggers from '../components/Debuggers';

export class Menu extends React.Component {
  static navigationOptions = {
    title: 'Settings',
  };

  constructor(props) {
    super(props);
    this._navigateToPasscodeForm = this._navigateToPasscodeForm.bind(this);
  }

  _navigateToPasscodeForm() {
    this.props.navigation.navigate('SetPasscode');
  }

  render() {
    return (
      <ScrollView>
        <Settings
          {...this.props}
          handleSetPasscode={this._navigateToPasscodeForm}
        />
        <Debuggers handlePasscodeReset={this.props.onPasscodeSettingsReset} />
      </ScrollView>
    );
  }
}

const mapStateToProps = state => ({
  requirePasscode: getPasscodeRequirement(state),
  currentPasscode: getCurrentPasscode(state),
});

export const mapDispatchToProps = dispatch => ({
  handleTogglePasscodeRequirement(currentPasscode, requirePasscode) {
    if (!currentPasscode && !requirePasscode) {
      Alert.alert(
        'No Passcode Set',
        'Please set passcode first before enabling this option',
        [
          {
            text: 'OK',
          },
        ],
        { cancelable: false }
      );
      return;
    }
    dispatch(togglePasscodeRequirement());
  },
  onPasscodeSettingsReset() {
    Alert.alert(
      'Reset Passcode Settings',
      'This will reset all your passcode settings to default (No passcode). This will not affect your entries.',
      [
        {
          text: 'Cancel',
          style: 'cancel',
        },
        {
          text: 'OK',
          onPress() {
            dispatch(resetPasscodeSettings());
          },
        },
      ],
      { cancelable: false }
    );
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Menu);
