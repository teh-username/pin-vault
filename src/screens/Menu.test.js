import React from 'react';
import { Alert } from 'react-native';

import { Menu, mapDispatchToProps as dispatchProps } from './Menu';
import {
  TOGGLE_PASSCODE_REQUIREMENT,
  RESET_PASSCODE_SETTINGS,
} from '../redux/modules/settings';

describe('Menu screen test', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<Menu />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('mapDispatchToProps test', () => {
    beforeEach(() => {
      jest.resetModules();
      jest.mock('Alert', () => ({
        alert: jest.fn(),
      }));
    });

    describe('handleTogglePasscodeRequirement test', () => {
      it('should not activate passscode requirement if user no passcode set yet', () => {
        const dispatchMock = jest.fn();
        const props = dispatchProps(dispatchMock);
        props.handleTogglePasscodeRequirement(null, false);
        expect(Alert.alert.mock.calls.length).toBe(1);
        expect(dispatchMock.mock.calls.length).toBe(0);
      });

      it('should activate passcode requirement if user has set a passcode', () => {
        const dispatchMock = jest.fn();
        const props = dispatchProps(dispatchMock);
        props.handleTogglePasscodeRequirement('HASH', true);
        expect(Alert.alert.mock.calls.length).toBe(0);
        expect(dispatchMock.mock.calls.length).toBe(1);
        expect(dispatchMock.mock.calls[0]).toEqual([
          { type: TOGGLE_PASSCODE_REQUIREMENT },
        ]);
      });
    });

    describe('onPasscodeSettingsReset test', () => {
      it('should dispatch reset action if user presses OK', () => {
        const dispatchMock = jest.fn();
        const props = dispatchProps(dispatchMock);
        props.onPasscodeSettingsReset();
        expect(Alert.alert.mock.calls.length).toBe(1);
        Alert.alert.mock.calls[0][2][1].onPress();
        expect(dispatchMock.mock.calls[0]).toEqual([
          { type: RESET_PASSCODE_SETTINGS },
        ]);
      });
    });
  });
});
