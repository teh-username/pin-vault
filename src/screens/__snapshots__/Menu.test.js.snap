// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu screen test should render as expected 1`] = `
<ScrollView>
  <Settings
    handleSetPasscode={[Function]}
  />
  <Debuggers />
</ScrollView>
`;
